etl/grapher_io.py

```diff
@@ -0,0 +1,796 @@
+) -> List[Dict[str, Any]]:
+    """Fetch data from S3 and add entity code and name from DB."""
+    args = [variable_ids]
+    if env:
+        args += [[env for _ in range(len(variable_ids))]]
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=workers) as executor:
+        results = list(executor.map(_fetch_metadata_from_s3, *args))
+
+    if not (isinstance(results, list) and all(isinstance(res, dict) for res in results)):
+        raise TypeError(f"results must be a list of dictionaries, got {type(results)}")
+
+    return results  # type: ignore
+
+
+def _fetch_metadata_from_s3(variable_id: int, env: OWIDEnv | None = None) -> Dict[str, Any] | None:
+    try:
+        # Cloudflare limits us to 600 requests per minute, retry in case we hit the limit
+        # NOTE: increase wait time or attempts if we hit the limit too often
+        for attempt in Retrying(
+            wait=wait_fixed(2),
+            stop=stop_after_attempt(3),
+            retry=retry_if_exception_type((URLError, RemoteDisconnected)),
+        ):
+            with attempt:
+                if env is not None:
+                    url = env.indicator_metadata_url(variable_id)
+                else:
+                    url = config.variable_metadata_url(variable_id)
+                return requests.get(url).json()
+    # no data on S3
+    except HTTPError:
+        return {}
+
+
+##############################################################################################
+# TO BE REVIEWED:
+# This is code that could be deprecated / removed?
+##############################################################################################
+
+
+def get_dataset_id(
+    dataset_name: str, db_conn: Optional[pymysql.Connection] = None, version: Optional[str] = None
+) -> Any:
+    """Get the dataset ID of a specific dataset name from database.
+
+    If more than one dataset is found for the same name, or if no dataset is found, an error is raised.
+
+    Parameters
+    ----------
+    dataset_name : str
+        Dataset name.
+    db_conn : pymysql.Connection
+        Connection to database. Defaults to None, in which case a default connection is created (uses etl.config).
+    version : str
+        ETL version of the dataset. This is necessary when multiple datasets have the same title. In such a case, if
+        version is not given, the function will raise an error.
+
+    Returns
+    -------
+    dataset_id : int
+        Dataset ID.
+
+    """
+    if db_conn is None:
+        db_conn = get_connection()
+
+    query = f"""
+        SELECT id
+        FROM datasets
+        WHERE name = '{dataset_name}'
+    """
+
+    if version:
+        query += f" AND version = '{version}'"
+
+    with db_conn.cursor() as cursor:
+        cursor.execute(query)
+        result = cursor.fetchall()
+
+    assert len(result) == 1, f"Ambiguous or unknown dataset name '{dataset_name}'"
+    dataset_id = result[0][0]
+    return dataset_id
+
+
+def get_variables_in_dataset(
```

Comment:
Eventually, I think we should move this as a method to `grapher_model.Dataset` so that all dataset-related methods are in one place. (Not now though!)